File: Engine/src/features.cpp
#include "features.hpp"
#include <algorithm>
#include <numeric>
#include <cmath>

namespace quant {

std::vector<double> sma(const std::vector<double>& prices, int period) {
    std::vector<double> result(prices.size(), std::nan(""));
    
    if (prices.size() < static_cast<size_t>(period)) {
        return result;
    }
    
    double sum = 0.0;
    for (int i = 0; i < period; ++i) {
        sum += prices[i];
    }
    result[period - 1] = sum / period;
    
    for (size_t i = period; i < prices.size(); ++i) {
        sum = sum - prices[i - period] + prices[i];
        result[i] = sum / period;
    }
    
    return result;
}

std::vector<double> ema(const std::vector<double>& prices, int period) {
    std::vector<double> result(prices.size(), std::nan(""));
    
    if (prices.size() < static_cast<size_t>(period)) {
        return result;
    }
    
    double alpha = 2.0 / (period + 1.0);
    double ema_value = prices[0];
    
    for (size_t i = 1; i < prices.size(); ++i) {
        ema_value = alpha * prices[i] + (1 - alpha) * ema_value;
        if (i >= static_cast<size_t>(period - 1)) {
            result[i] = ema_value;
        }
    }
    
    return result;
}

std::vector<double> rsi(const std::vector<double>& prices, int period) {
    std::vector<double> result(prices.size(), std::nan(""));
    
    if (prices.size() < static_cast<size_t>(period + 1)) {
        return result;
    }
    
    std::vector<double> gains, losses;
    for (size_t i = 1; i < prices.size(); ++i) {
        double change = prices[i] - prices[i - 1];
        gains.push_back(change > 0 ? change : 0.0);
        losses.push_back(change < 0 ? -change : 0.0);
    }
    
    double avg_gain = 0.0, avg_loss = 0.0;
    for (int i = 0; i < period; ++i) {
        avg_gain += gains[i];
        avg_loss += losses[i];
    }
    avg_gain /= period;
    avg_loss /= period;
    
    if (avg_loss == 0.0) {
        result[period] = 100.0;
    } else {
        double rs = avg_gain / avg_loss;
        result[period] = 100.0 - (100.0 / (1.0 + rs));
    }
    
    // Wilder's smoothing
    double alpha = 1.0 / period;
    for (size_t i = period + 1; i < prices.size(); ++i) {
        avg_gain = alpha * gains[i - 1] + (1 - alpha) * avg_gain;
        avg_loss = alpha * losses[i - 1] + (1 - alpha) * avg_loss;
        
        if (avg_loss == 0.0) {
            result[i] = 100.0;
        } else {
            double rs = avg_gain / avg_loss;
            result[i] = 100.0 - (100.0 / (1.0 + rs));
        }
    }
    
    return result;
}

std::vector<double> atr(
    const std::vector<double>& high,
    const std::vector<double>& low,
    const std::vector<double>& close,
    int period) {
    
    std::vector<double> result(close.size(), std::nan(""));
    
    if (high.size() != low.size() || high.size() != close.size() ||
        high.size() < static_cast<size_t>(period + 1)) {
        return result;
    }
    
    std::vector<double> true_ranges;
    for (size_t i = 1; i < high.size(); ++i) {
        double tr1 = high[i] - low[i];
        double tr2 = std::abs(high[i] - close[i - 1]);
        double tr3 = std::abs(low[i] - close[i - 1]);
        true_ranges.push_back(std::max({tr1, tr2, tr3}));
    }
    
    // Calculate initial ATR
    double atr_value = 0.0;
    for (int i = 0; i < period; ++i) {
        atr_value += true_ranges[i];
    }
    atr_value /= period;
    result[period] = atr_value;
    
    // Wilder's smoothing
    double alpha = 1.0 / period;
    for (size_t i = period; i < true_ranges.size(); ++i) {
        atr_value = alpha * true_ranges[i] + (1 - alpha) * atr_value;
        result[i + 1] = atr_value;
    }
    
    return result;
}

std::vector<double> momentum(const std::vector<double>& prices, int period) {
    std::vector<double> result(prices.size(), std::nan(""));
    
    if (prices.size() < static_cast<size_t>(period + 1)) {
        return result;
    }
    
    for (size_t i = period; i < prices.size(); ++i) {
        if (prices[i - period] != 0.0) {
            result[i] = (prices[i] - prices[i - period]) / prices[i - period];
        }
    }
    
    return result;
}

std::vector<double> rolling_std(const std::vector<double>& values, int period) {
    std::vector<double> result(values.size(), std::nan(""));
    
    if (values.size() < static_cast<size_t>(period)) {
        return result;
    }
    
    for (size_t i = period - 1; i < values.size(); ++i) {
        double sum = 0.0;
        double sum_sq = 0.0;
        
        for (int j = 0; j < period; ++j) {
            double val = values[i - j];
            sum += val;
            sum_sq += val * val;
        }
        
        double mean = sum / period;
        double variance = (sum_sq / period) - (mean * mean);
        result[i] = std::sqrt(variance);
    }
    
    return result;
}

std::vector<double> zscore(
    const std::vector<double>& values,
    int period) {
    
    std::vector<double> result(values.size(), std::nan(""));
    
    if (values.size() < static_cast<size_t>(period)) {
        return result;
    }
    
    for (size_t i = period - 1; i < values.size(); ++i) {
        double sum = 0.0;
        double sum_sq = 0.0;
        
        for (int j = 0; j < period; ++j) {
            double val = values[i - j];
            sum += val;
            sum_sq += val * val;
        }
        
        double mean = sum / period;
        double variance = (sum_sq / period) - (mean * mean);
        double std_dev = std::sqrt(variance);
        
        if (std_dev > 0.0) {
            result[i] = (values[i] - mean) / std_dev;
        }
    }
    
    return result;
}

} // namespace quant

File: Research/src/features.py
import numpy as np
import pandas as pd

def add_log_returns(df: pd.DataFrame, target_col: str = "close") -> pd.DataFrame:
    """
    Calculates log returns.
    Definition: r_t = ln(P_t / P_{t-1})
    """
    df = df.copy()
    df['log_return'] = np.log(df[target_col] / df[target_col].shift(1))
    return df

def add_rolling_volatility(df: pd.DataFrame, log_ret_col: str = "log_return") -> pd.DataFrame:
    """
    Adds rolling volatility (standard deviation of log returns).
    """
    df = df.copy()
    df['vol_20'] = df[log_ret_col].rolling(20).std()
    df['vol_60'] = df[log_ret_col].rolling(60).std()
    
    # Fill zeros or NaNs to avoid division errors later
    df['vol_20'] = df['vol_20'].replace(0, np.nan).ffill()
    df['vol_60'] = df['vol_60'].replace(0, np.nan).ffill()
    return df

def add_moving_averages(df: pd.DataFrame, price_col: str = "close") -> pd.DataFrame:
    """
    Adds Simple Moving Averages (SMA).
    """
    df = df.copy()
    df['sma_20'] = df[price_col].rolling(window=20).mean()
    df['sma_60'] = df[price_col].rolling(window=60).mean()
    return df

def add_ema(df: pd.DataFrame, periods: list = [12, 26], price_col: str = "close") -> pd.DataFrame:
    """
    Adds Exponential Moving Averages (EMA).
    """
    df = df.copy()
    for period in periods:
        df[f'ema_{period}'] = df[price_col].ewm(span=period, adjust=False).mean()
    return df

def add_rsi(df: pd.DataFrame, period: int = 14, price_col: str = "close") -> pd.DataFrame:
    """
    Adds Relative Strength Index (RSI).
    """
    df = df.copy()
    delta = df[price_col].diff()
    
    gain = delta.where(delta > 0, 0.0)
    loss = (-delta).where(delta < 0, 0.0)
    
    avg_gain = gain.ewm(alpha=1/period, min_periods=period, adjust=False).mean()
    avg_loss = loss.ewm(alpha=1/period, min_periods=period, adjust=False).mean()
    
    rs = avg_gain / avg_loss
    df[f'rsi_{period}'] = 100 - (100 / (1 + rs))
    df[f'rsi_{period}'] = df[f'rsi_{period}'].replace([np.inf, -np.inf], np.nan).ffill()
    return df

def add_macd(df: pd.DataFrame, fast: int = 12, slow: int = 26, signal: int = 9, price_col: str = "close") -> pd.DataFrame:
    """
    Adds MACD (Line, Signal, Histogram).
    """
    df = df.copy()
    ema_fast = df[price_col].ewm(span=fast, adjust=False).mean()
    ema_slow = df[price_col].ewm(span=slow, adjust=False).mean()
    
    df['macd_line'] = ema_fast - ema_slow
    df['macd_signal'] = df['macd_line'].ewm(span=signal, adjust=False).mean()
    df['macd_histogram'] = df['macd_line'] - df['macd_signal']
    return df

def add_bollinger_bands(df: pd.DataFrame, period: int = 20, num_std: float = 2.0, price_col: str = "close") -> pd.DataFrame:
    """
    Adds Bollinger Bands (Upper, Lower, Middle, Width, Position).
    """
    df = df.copy()
    df['bb_middle'] = df[price_col].rolling(period).mean()
    rolling_std = df[price_col].rolling(period).std()
    
    df['bb_upper'] = df['bb_middle'] + (num_std * rolling_std)
    df['bb_lower'] = df['bb_middle'] - (num_std * rolling_std)
    
    # Bandwidth: volatility measure
    df['bb_width'] = (df['bb_upper'] - df['bb_lower']) / df['bb_middle']
    
    # Position: -1 at lower band, 0 at middle, +1 at upper band
    df['bb_position'] = (df[price_col] - df['bb_middle']) / (num_std * rolling_std)
    return df

def add_atr(df: pd.DataFrame, period: int = 14) -> pd.DataFrame:
    """
    Adds Average True Range (ATR).
    """
    df = df.copy()
    high_low = df['high'] - df['low']
    high_close = (df['high'] - df['close'].shift()).abs()
    low_close = (df['low'] - df['close'].shift()).abs()
    
    true_range = pd.concat([high_low, high_close, low_close], axis=1).max(axis=1)
    df[f'atr_{period}'] = true_range.ewm(alpha=1/period, min_periods=period, adjust=False).mean()
    return df

def add_momentum(df: pd.DataFrame, periods: list = [10, 20], price_col: str = "close") -> pd.DataFrame:
    """
    Adds Price Momentum and Momentum Z-Score.
    """
    df = df.copy()
    for period in periods:
        # Simple momentum (rate of change)
        df[f'momentum_{period}'] = df[price_col].pct_change(period)
        
        # Momentum z-score (standardized)
        rolling_mean = df[f'momentum_{period}'].rolling(period).mean()
        rolling_std = df[f'momentum_{period}'].rolling(period).std()
        df[f'momentum_zscore_{period}'] = (df[f'momentum_{period}'] - rolling_mean) / rolling_std
        
        df[f'momentum_zscore_{period}'] = df[f'momentum_zscore_{period}'].replace([np.inf, -np.inf], np.nan).ffill()
    return df

def add_return_zscore(df: pd.DataFrame) -> pd.DataFrame:
    """
    Adds Z-Score based on Log Returns (from original file 1).
    Requires 'log_return' and 'vol_20' to exist.
    """
    df = df.copy()
    mean_20 = df['log_return'].rolling(20).mean()
    
    # Standard Z-Score formula: (Value - Mean) / StdDev
    df['zscore_20'] = (df['log_return'] - mean_20) / df['vol_20']
    
    df['zscore_20'] = df['zscore_20'].replace([np.inf, -np.inf], np.nan).ffill()
    return df

def generate_features(df: pd.DataFrame) -> pd.DataFrame:
    """
    MASTER FUNCTION: Runs the entire feature engineering pipeline.
    """
    # 1. Base Calculations
    df = add_log_returns(df)
    df = add_rolling_volatility(df)
    
    # 2. Trends & Averages
    df = add_moving_averages(df)
    df = add_ema(df, periods=[12, 26])
    
    # 3. Oscillators & Momentum
    df = add_rsi(df, period=14)
    df = add_macd(df)
    df = add_momentum(df, periods=[10, 20])
    
    # 4. Volatility Bands & ATR
    df = add_bollinger_bands(df, period=20, num_std=2.0)
    df = add_atr(df, period=14)
    
    # 5. Advanced Stats (Requires previous steps)
    df = add_return_zscore(df)
    
    # 6. Final Cleanup
    # Drop initial rows where rolling windows haven't filled yet
    df = df.dropna()
    
    return df

# USAGE:
# from features import generate_features
# df_enriched = generate_features(df_raw)